### PR TITLE
Transformer experiments

### PR DESCRIPTION
**Pull request by garrettbingham**
Main changes:
* Support for new layer types 
* Support for API calls like tf.nn.relu that get wrapped in TfOpLambda layers
* Support for signal variance other than 1.0
* Other minor stuff

Pylint is happy

### CODE REVIEW

File: autoinit/components/weighted_sum.py [hunk 1]
```diff
@@ -23,27 +23,22 @@ class WeightedSum(tfkeras.layers.Layer):
...
+    def __init__(self, coefficients=None, **kwargs):
         """
         The constructor initializes the arguments and the
         base class.
         :param num_inputs: Number of inputs to this Layer
         """
         super().__init__(**kwargs)
-        assert num_inputs >= 2
-        self.num_inputs = num_inputs
+        self.coefficients = coefficients
```
Comment (garrettbingham) [inline autoinit/components/weighted_sum.py:33]: The coefficients need to be preserved as state, not used as weights.  With state, the same coefficients can be maintained when the layer is serialized/deserialized.  With weights, the weights were being forgotten and were re-initialized as ones.

File: autoinit/estimators/activation.py [hunk 2]
```diff
@@ -75,11 +76,20 @@ def estimate(self, means_in: List, vars_in: List):
...
-            # We can't integrate over softmax.  At initialization, we expect
-            # balanced logits with mean 1 / NUM_CLASSES and variance 1 / NUM_CLASSES².
-            num_classes = prod(self.layer.output.shape[1:])
-            mean_out = 1.0 / num_classes
-            var_out = 1.0 / math.pow(num_classes, 2)
+            # We can't integrate over softmax, so we use Monte Carlo sampling.
+            if hasattr(self.layer, 'axis'):
+                num_classes = np.prod(self.layer.input_shape[self.layer.axis])
+            else:
+                num_classes = np.prod(self.layer.output.shape[1:])
```
Comment (garrettbingham) [inline autoinit/estimators/activation.py:83]: Check for the axis attribute in case this softmax is part of an attention layer

File: autoinit/estimators/dense.py [hunk 3]
```diff
@@ -87,36 +120,46 @@ def estimate(self, means_in: List, vars_in: List):
...
+        scale = self.signal_variance / second_moment
         gain = math.sqrt(scale)
-        distribution = self.estimator_config.get("distribution",
-                                                 "truncated_normal")
+        distribution = self.estimator_config.get("distribution", "truncated_normal")
+        if isinstance(self.layer, DEPTHWISE_CONV):
+            # For depthwise convolutional layers, each channel is convolved with a different
+            # kernel, so we need to multiply the scale/gain by the number of channels.
+            scale *= self.layer.input.shape[-1]
+            gain *= self.layer.input.shape[-1]
```
Comment (garrettbingham) [inline autoinit/estimators/dense.py]: depthwise conv layers need special treatment

File: autoinit/estimators/estimate_layer_output_distribution.py [hunk 4]
```diff
@@ -31,15 +31,21 @@ class LayerOutputDistributionEstimator:
...
         This constructor initializes the params.
         :param layer: TensorFlow/Keras Layer
         :param estimator_config: Dictionary containing the required config for
         layer.
+        :param signal_variance: Signal variance to maintain. This parameter is useful if a value
+            other than the default variance = 1.0 is desired.
         """
         self.layer = layer
         self.estimator_config = estimator_config
+        self.signal_variance = signal_variance
```
Comment (garrettbingham) [inline autoinit/estimators/estimate_layer_output_distribution.py:48]: another important change; comment says what it does

File: autoinit/estimators/tf_op_lambda.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+class TFOpLambdaOutputDistributionEstimator(LayerOutputDistributionEstimator):
+    """
+    TensorFlow wraps API calls (like tf.nn.relu) in a TFOpLambda Layer object.
+    Similarly, sometimes users wrap arbitrary functions in Lambda layers.
+    In these cases, the best we can do is to try to infer the layer operation
+    from the layer name.  Often the layer names will have suffixes, like tf.reshape_3,
+    so we just check whether these API calls are in the layer name.  This class is a
+    workaround.  If possible, API calls should be replaced with proper Layer objects.
+    """
```
Comment (garrettbingham) [inline autoinit/estimators/tf_op_lambda.py:36]: This whole implementation is a bit of a hack, but with API calls we don't have access to the usual Layer attributes.  Inferring functionality from the name seems like the least bad approach.

File: setup.py [hunk 6]
```diff
@@ -12,7 +12,7 @@
 from setuptools import setup, find_packages
 
 setup(name='autoinit',
-    version='0.0.5',
+    version='1.1.0',
```
Comment (garrettbingham) [inline setup.py:15]: Current release is 1.0.0

File: autoinit/estimators/dense.py [hunk 7]
```diff
@@ -39,13 +40,45 @@
...
 
+DENSE = tfkeras.layers.Dense
+CONV_1D = (tfkeras.layers.Conv1D, tfkeras.layers.DepthwiseConv1D)
+CONV_2D = (tfkeras.layers.Conv2D, tfkeras.layers.DepthwiseConv2D)
+CONV_3D = tfkeras.layers.Conv3D
+
+DENSE_OR_CONV = (tfkeras.layers.Dense,
+                 tfkeras.layers.Conv1D,
+                 tfkeras.layers.Conv2D,
+                 tfkeras.layers.Conv3D)
```
Comment (ofrancon) [inline autoinit/estimators/dense.py]: What about `tfkeras.layers.DepthwiseConv1D`? Don't you want it to be considered CONV too?
If you do then you should do:
```suggestion
DENSE_OR_CONV = (DENSE,
                CONV_1D,
                CONV_2D,
                CONV_3D)
```

File: autoinit/estimators/tf_op_lambda.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+        return mean_out, var_out
+
+    def _relu(self, means_in, vars_in):
+        # See activation.py
+        return self._mapped_distribution(tfkeras.activations.relu, means_in[0], vars_in[0])
+
+    def _get_tf_op_lambda_estimators(self):
+        # Map layer names to estimators.
+        return {
```
Comment (ofrancon) [inline autoinit/estimators/tf_op_lambda.py]: A new map is going to be constructed each time you call this function. If performance is a problem, it could be a class variable instead. Maybe even a constant, as the mapping won't change from instance to instance.

File: autoinit/initializer/output_distribution_factory.py [hunk 9]
```diff
@@ -59,11 +65,16 @@ class OutputDistributionEstimatorFactory:
...
+                 custom_distribution_estimators: Dict = None,
+                 signal_variance: float = 1.0):
         """
         The constructor consists of the supported layer types.
         """
-        custom_distribution_estimators = custom_distribution_estimators or {}
+        self.estimator_config = estimator_config
+        self.signal_variance = signal_variance
+
         self.builtin_layers = {
```
Comment (ofrancon) [inline autoinit/initializer/output_distribution_factory.py:78]: Could be a constant

**Review by ofrancon — APPROVED** [review]
Minor comments, lgtm!

File: autoinit/estimators/activation.py [hunk 10]
```diff
@@ -75,11 +76,20 @@ def estimate(self, means_in: List, vars_in: List):
...
+            num_samples = self.estimator_config.get("monte_carlo_samples",
+                                                    DEFAULT_MONTE_CARLO_SAMPLES)
+            samples = np.random.normal(loc=means_in[0],
+                                       scale=np.sqrt(vars_in[0]),
+                                       size=(int(num_samples), num_classes))
+            out = np.exp(samples) / np.sum(np.exp(samples), axis=1, keepdims=True)
```
Comment (EKMeyerson) [inline autoinit/estimators/activation.py:89]: Nice mcmc here. Do you have an idea of how/whether the number of samples needed for a decent variance estimate changes with the size of the softmax layer?

File: autoinit/estimators/dense.py [hunk 11]
```diff
@@ -87,36 +120,46 @@ def estimate(self, means_in: List, vars_in: List):
...
         second_moment = vars_in[0] + math.pow(means_in[0], 2)
-        scale = 1.0 / second_moment
+        scale = self.signal_variance / second_moment
         gain = math.sqrt(scale)
-        distribution = self.estimator_config.get("distribution",
-                                                 "truncated_normal")
+        distribution = self.estimator_config.get("distribution", "truncated_normal")
+        if isinstance(self.layer, DEPTHWISE_CONV):
+            # For depthwise convolutional layers, each channel is convolved with a different
+            # kernel, so we need to multiply the scale/gain by the number of channels.
```
Comment (EKMeyerson) [inline autoinit/estimators/dense.py]: Took a bit to figure out this comment, maybe more clear:

"For depthwise convolutional layers, each kernel takes a single channel as input (instead of all channels), so we need to multiply the scale/gain by the number of channels to recover the intended output variance."

File: autoinit/estimators/tf_op_lambda.py (hunk 5 above)
Comment (EKMeyerson) [reply autoinit/estimators/tf_op_lambda.py:36]: Why didn't you need this until now?

File: autoinit/estimators/tf_op_lambda.py (hunk 8 above)
Comment (EKMeyerson) [reply autoinit/estimators/tf_op_lambda.py]: +1 for constant

**Review by EKMeyerson — APPROVED** [review]
minor comments/questions; lgtm o/w

File: autoinit/estimators/tf_op_lambda.py (hunk 5 above)
Comment (garrettbingham) [reply autoinit/estimators/tf_op_lambda.py:36]: The architectures I was using only used Layers.  When I applied AutoInit to CoAtNet there were a lot of these API calls.

File: autoinit/estimators/dense.py (hunk 7 above)
Comment (garrettbingham) [reply autoinit/estimators/dense.py]: No, I'm just checking whether or not it is a depthwise conv because in that case we need to set `depthwise_initializer` instead of `kernel_initializer`.  Good point though; the code is ambiguous.  I'll make it more clear.

Comment (garrettbingham) [reply autoinit/estimators/dense.py]: Renamed to `DENSE_OR_REGULAR_CONV`

File: autoinit/estimators/tf_op_lambda.py (hunk 8 above)
Comment (garrettbingham) [reply autoinit/estimators/tf_op_lambda.py]: Good call; fixed.

File: autoinit/estimators/tf_op_lambda.py (hunk 5 above)
Comment (EKMeyerson) [reply autoinit/estimators/tf_op_lambda.py:36]: makes sense; thanks

File: autoinit/estimators/activation.py (hunk 10 above)
Comment (garrettbingham) [reply autoinit/estimators/activation.py:89]: Not sure. In experiments 1e5 seemed fast enough and accurate enough.

File: autoinit/estimators/dense.py (hunk 11 above)
Comment (garrettbingham) [reply autoinit/estimators/dense.py]: Yes, much better, thanks!
